nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/CoreSession.java

```diff
@@ -1438,4 +1438,15 @@ public static boolean isResetCreator(CopyOption... options) {
      */
     String getChangeToken(DocumentRef ref);
 
+    /**
+     * Gets a document if it exists, otherwise creates it. This is done atomically to prevent different threads from
+     * trying to create the same document. If the document did not exist and is therefore created, the current
+     * transaction is committed and the newly-created document is also committed in its own transaction
+     *
+     * @param docModel the document model
+     * @return the existing or created document
+     * @since 9.3
+     */
+    DocumentModel getOrCreateDocument(DocumentModel docModel);
```

Comment:
ok for the javadoc but the method name don't reflect the behavior, using this to replace a createDocument will results in a complex bug in case of rollback, I prefer something like getOrCommitAndCreateDocument or anything that reflects that the current transaction is impacted.

Reply:
I prefer to put a big warning in the Javadoc rather than changing the method name.